src/contrib/tf_op/tvm_dso_op_kernels.cc

```diff
@@ -0,0 +1,285 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <cstdio>
+#include <cuda_runtime.h>
+#include <dlpack/dlpack.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/packed_func.h>
+#include "tensorflow/core/framework/op_kernel.h"
+
+#include "index_seq.h"
+
+using namespace tensorflow;
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+typedef Eigen::GpuDevice GPUDevice;
+typedef gtl::InlinedVector<int64, 4> ShapeContainer;
+
+
+template <typename DEVICE_TYPE>
+class TVMDSOOpTrait;
+
+
+class TensorAsBuf {
+  public:
+    Tensor inline_tensor;
+    Tensor* tensor;
+
+    size_t size;
+    size_t offset;
+
+    int device_type;
+
+    char* origin_buf; 
+    char* buf;
+
+    void CopyToOrigin() {
+        if (buf == origin_buf) {
+            return;
+        }
+        if (device_type == kDLCPU) {
+            memcpy(origin_buf, buf + offset, size); 
+        } else {
```

Comment:
`else if (device_type == kDLGPU) {cudaMemcpy...} else { ...` We have many device type, not only cpu and cuda.

Reply:
Now we only support CPU and CUDA devices. We can left the TODO comment if we support other device in the future.

Reply:
It is not a good idea. For example, when we set the context be OpenCL, what happened? we will go into `else`. It is no problem only support CPU and CUDA devices now, but you could implement like this:
```
if (device_type == kDLCPU) {

} else if (device_type == kDLGPU) {

} else {
   LOG(FATAL) << "Only support CPU and CUDA now. Not implemented " << device_type << "currently"; 
}